src/system/random/basic.lean

```diff
@@ -0,0 +1,467 @@
+
+/-- use `random` in the `tactic` monad -/
+meta def random : tactic α :=
+run_rand (rand.random α)
+
+/-- use `random_series` in the `tactic` monad -/
+meta def random_series : tactic (stream α) :=
+run_rand (rand.random_series α)
+
+end random
+
+end tactic
+
+namespace bool
+
+/-- Make `i` into an element of the interval `x .. y` if feasible
+and return an arbitrary element of `x .. y` otherwise -/
+def fit_in_Icc (x y : bool) (p : x ≤ y) (i : bool) : x .. y := do
+  if hx : x ≤ i ∧ i ≤ y
+  then ⟨ i, hx ⟩
+  else ⟨ x , le_refl x , p ⟩
+
+open ulift (hiding inhabited)
+variables {g : Type} [random_gen g]
+
+/-- generate a randomly generated boolean value -/
+protected def get_random : rand_g g bool :=
+⟨ prod.map id up ∘ @rand_bool g _ ∘ down ⟩
+
+end bool
+
+instance : random bool :=
+{ random   := λ g, @bool.get_random _,
+  random_r := λ g _inst x y p, bool.fit_in_Icc _ _ p <$> (@bool.get_random g _inst) }
+
+open nat (succ one_add mod_eq_of_lt zero_lt_succ add_one succ_le_succ)
+
+namespace stream
+
+variable {α : Type u}
+
+open list (length) stream (approx)
+
+lemma length_approx :
+  ∀ (s : stream α) (n : ℕ), length (approx n s) = n
+| s 0 := rfl
+| s (succ n) := by simp [approx,length,one_add,length_approx]
+
+end stream
+
+variables {g : Type} [random_gen g]
+
+/-- generate a random bit vector of length `n` -/
+def bitvec.random (n : ℕ) : rand_g g (bitvec n) := do
+bs ← rand.random_series bool,
+pure ⟨bs.take n, bs.length_take _⟩
+
+section fit_in_Icc
+
+parameters {i' n : ℕ}
+parameters {x y : bitvec n}
+
+parameters P' : x.to_nat ≤ y.to_nat
+include P'
+
+local infix ^ := nat.pow
+
+lemma bitvec.interval_fits_in_word_size :
+  x.to_nat + i' % (1 + (y.to_nat - x.to_nat)) < 2^n :=
+begin
+  let x' := x.to_nat,
+  let y' := y.to_nat,
+  apply @lt_of_lt_of_le _ _ _ (x' + (y' - x' + 1)),
+  { apply add_lt_add_left, simp only [add_comm _ 1],
+    apply nat.mod_lt,
+    simp [add_comm 1,zero_lt_succ], },
+  { rw [← add_assoc,← nat.add_sub_assoc P',nat.add_sub_cancel_left,add_one],
+    clear P' i',
+    cases y with y Hy,
+    unfold bitvec.to_nat vector.to_list subtype.val bitvec.bits_to_nat,
+    rw [← reverse_reverse y, foldl_reverse,← Hy,← length_reverse],
+    generalize : reverse y = z,
+    clear x' x y' Hy y n,
+    induction z with x xs,
+    { rw [list.length,list.foldr,nat.pow] },
+    { simp [foldr,length,one_add,pow_succ,flip,bitvec.add_lsb,add_comm _ 1],
+      transitivity succ
+       (foldr (λ (b : bool) (a : ℕ), a + a + cond b 1 0) 0 xs +
+          foldr (λ (b : bool) (a : ℕ), a + a + cond b 1 0) 0 xs + 1),
+      { apply succ_le_succ, apply add_le_add_left,
+        cases x, apply nat.zero_le, refl, },
+      { simp! only,
+        rw [← nat.succ_eq_add_one,← nat.succ_add,← nat.add_succ],
+        rw [mul_comm _ 2,← two_mul],
+        apply nat.mul_le_mul_left,
+        simpa [flip,bitvec.add_lsb] using z_ih, } }, },
+end
+end fit_in_Icc
+
+open nat
+
+/-- Use `i` to generate an element of the interval `x .. y` -/
```

Comment:
We should warn here that the result will not be uniformly distributed if `y - x + 1` does not divide `2^n` (for example, `x = 0b00` and `y = 0b10` results in this map: `0b00 -> 0b00, 0b01 -> 0b01, 0b10 -> 0b10, 0b11 -> 0b00`).